datafusion/optimizer/src/push_down_limit.rs

```diff
@@ -238,11 +218,57 @@ impl OptimizerRule for PushDownLimit {
     }
 }
 
-fn fetch_minus_skip(fetch: usize, skip: usize) -> usize {
-    if skip > fetch {
-        0
+fn push_down_join(join: &Join, limit: usize) -> Option<Join> {
+    use JoinType::*;
+
+    fn is_no_join_condition(join: &Join) -> bool {
+        join.on.is_empty() && join.filter.is_none()
+    }
+
+    let (left_limit, right_limit) = if is_no_join_condition(join) {
+        match join.join_type {
+            Left | Right | Full => (Some(limit), Some(limit)),
+            LeftAnti | LeftSemi => (Some(limit), None),
+            RightAnti | RightSemi => (None, Some(limit)),
+            Inner => (None, None),
+        }
     } else {
-        fetch - skip
+        match join.join_type {
+            Left => (Some(limit), None),
+            Right => (None, Some(limit)),
+            _ => (None, None),
+        }
+    };
+
+    if let (None, None) = (left_limit, right_limit) {
```

Comment:
I think this would be more clear with `match` here than using `if let`

Reply:
Updated.